ONSdigital/census-contact-centre-service
# Cr 748

# Motivation and Context

This change is required because the FulfilmentRequest event, which is generated by the CCSVC when it receives a 'fulfilmentRequestByPost' request, needs to include Address details, 
which are currently empty.

It has not yet been decided (by the business) whether the address details, which are sent to RM in the FulfilmentRequest event, should come from the RM case or, if the address has been modified 
but not yet validated, should be the modified ones. Currently the address is always taken from the RM case.

# How to test?
These changes can be tested using postman by doing as follows:
1. Run rabbitmq, mock case service, and this branch of the contact centre service.
2. Go to the rabbitmq management console and purge the queue, named event.fulfilment.request, to make sure that it is empty: http://localhost:46672/#/queues/%2F/event.fulfilment.request
3. Add the following JSON request to the mock case service using this endpoint: http://localhost:8161/cases/data/cases/add

    [{
        "caseRef": "123123",
        "arid": "2344266233",
        "estabArid": "AABBCC",
        "estabType": "ET",
        "uprn": "1347459999",
        "createdDateTime": "2019-04-14T13:45:26.564+01:00",
        "addressLine1": "Napier House",
        "addressLine2": "88 Harbour Street",
        "addressLine3": "Parkhead",
        "townName": "Glasgow",
        "postcode": "G1 2AA",
        "organisationName": "ON",
        "addressLevel": "E",
        "abpCode": "AACC",
        "latitude": "41.40338",
        "longitude": "2.17403",
        "oa": "EE22",
        "lsoa": "x1",
        "msoa": "x2",
        "lad": "H1",
        "caseEvents": [
            {
                "id": "101",
                "eventType": "CASE_UPDATED",
                "description": "Initial creation of case",
                "createdDateTime": "2019-04-14T13:45:26.564+01:00"
            },
            {
                "id": "102",
                "eventType": null,
                "description": "Create Household Visit",
                "createdDateTime": "2019-04-14T13:45:26.564+01:00"
            }
        ],
        "id": "9b90abb2-6105-4d0d-9dc9-1054b61a29f6",
        "caseType": "HH",
        "region": "E",
        "collectionExerciseId": "49871667-117d-4a63-9101-f6a0660f73f6",
        "surveyType": "CENSUS",
        "handDelivery": false
    }] 

4. Go to the following POST endpoint: http://localhost:8171/cases/9b90abb2-6105-4d0d-9dc9-1054b61a29f6/fulfilment/post
5. From there send the following request body:

{
	"caseId": "9b90abb2-6105-4d0d-9dc9-1054b61a29f6",
	"title": "Mrs",
	"forename": "Sally",
	"surname": "Smurf",
	"fulfilmentCode": "P_OR_H1",
	"dateTime": "2019-04-14T13:45:26.564+01:00"
} 

6. Go back to the rabbitmq management console and make sure that there is now 1 message in the queue named event.fulfilment.request
7. Check that the message is a FULFILMENT_REQUESTED event that contains all the address fields as expected:
{"event":{"type":"FULFILMENT_REQUESTED","source":"CONTACT_CENTRE_API","channel":"CC","dateTime":"2020-03-11T13:12:10.975Z","transactionId":"dd8acba8-006c-42f5-ad05-5208045ccdfd"},"payload":{"fulfilmentRequest":{"fulfilmentCode":"P_OR_H1","caseId":"3305e937-6fb1-4ce1-9d4c-077f147789aa","individualCaseId":null,"address":{"addressLine1":"Napier House","addressLine2":"88 Harbour Street","addressLine3":"Parkhead","townName":"Glasgow","postcode":"G1 2AA","region":"E","latitude":"41.40338","longitude":"2.17403","uprn":"1347459999","arid":"2344266233","addressType":null,"estabType":"ET"},"contact":{"title":"Mrs","forename":"Sally","surname":"Smurf","email":null,"telNo":null}}}}
